scylla/src/transport/connection.rs

```diff
@@ -649,22 +658,35 @@ impl Connection {
                     "Connection broken",
                 )))
             })?;
-
-        let task_response = tokio::time::timeout(self.config.client_timeout, receiver)
+        let stream_id = stream_id_receiver
             .await
-            .map_err(|e| {
-                QueryError::ClientTimeout(format!(
+            .map_err(|_| QueryError::UnableToAllocStreamId)?;
+        let received = match tokio::time::timeout(self.config.client_timeout, receiver).await {
```

Comment:
Library user can decide to stop polling future returned by `Session::query*` method, causing this line to be last line executed. In this case no `Task::Orphan` message will be send via `self.submit_channel`.

I think that speculative execution can also cancel futures, so the problem persists even without user intentionally canceling futures.

Reply:
It's true, but it's also no worse than what we have now - the connection would eventually go out of stream ids and die. Users should be encouraged to use our timeout instead of tokio timeout. Do you have any suggestions on how to approach this issue?

Reply:
One idea that comes to my mind would be to make the `router` responsible for detecting timeouts and marking stream IDs as orphaned. It would keep a `BTreeMap`-based queue of stream IDs ordered by their deadlines.

Reply:
I'm more for a different solution.
The problem in current solution is that we are not able to deliver notification about orphanage to the `router`, due to future cancellation.

To ensure, that some action is run in an async function, even when a function's future was dropped, one can use a destructor of an object living in the body of this function. I propose introducing such object to the `send_request` function, to use it for notifying `router` about orphanage.

Due to destructors not being async, notifying via bounded channel as it was done now is not an option. Luckily, unbounded tokio channels do not require awaiting to send some value, so they can be used in rust's drop implementations.



```rust
struct OrphanhoodNotifier {
    disabled: bool, // initialized to false
    stream_id: i16,
    sender: UnboundedSender<i16>,
}

impl OrphanhoodNotifier {
    fn disable(mut self) {
        self.disabled = true;
    }
}

impl Drop for OrphanhoodNotifier {
    fn drop(&mut self) {
        if (!self.disabled) {
            self.sender.send(self.stream_id);
        }
    }
}
```

I imagine, that the `OrphanhoodNotifier` would be sent instead of raw `stream_id` using `stream_id_sender` (as this would protect us from not notifying about orphanage, when `send_request` is canceled before receiving `stream_id` from `stream_id_receiver` but after it submitted `Task::Request`). After receival of query response, orphan notification would be disabled by calling `.disable()` method of `OrphanhoodNotifier`.

Such design would also require changes in `router`'s way of allocating `stream_id`s and marking `stream_id`s as orphaned.
Why? Let's consider a situation, when:
1. A request is sent by `send_request`. Let's call this invocation of `send_request` "the first invocation".
2. The request sent by the first invocation gets a `stream_id = 1`.
3. Connection's `router` executes this request and, after receiving response, `stream_id = 1` goes back to pool.
4. Next, another request is sent by another invocation of `send_request` (let's call it "the second invocation").
5. The second request also gets a `stream_id = 1`, because it was already available in a pool.
6. Now, if a late orphanage notification somehow arrives from the first invocation of `send_request`, it would mark the wrong request (request currently having `stream_id = 1` sent by the `send_request` second invocation).

One way to get rid of this problem, is to track generations of stream_ids.
```rust
struct OrphanhoodNotifier {
    disabled: bool, // initialized to false
    stream_id: i16,
    generation: u64, // new field denoting the stream_id's generation number
    sender: UnboundedSender<i16>,
}
```
On the router side, an associative array containing mapping `stream_id` -> `current generation number` would be kept. When a new stream_id is allocated, it's generation number inside the associative array is incremented. If a orphanhood notification is received, there should be check if the generations match (the one in notification and other in associative array). After check succeeds, `stream_id` can be considered as orphaned. If the check does not succeed, we can be sure, that the router already received the response, and the `stream_id` was reused.


I believe this design is better than keeping BTreeMap queue of stream IDs, because it allows to handle timeouts by future cancellation (and therefore allows us to implement timeouts without knowing the timeout duration in advance). Being able to handle cancellations is crucial as the implementation of speculative execution heavily depends on them (by using `select!`).